rfd/0151-database-permission-management.md

```diff
@@ -0,0 +1,410 @@
+---
+authors: Krzysztof Skrzętnicki (krzysztof.skrzetnicki@goteleport.com)
+state: draft
+---
+
+# RFD 151 - Database Permission Management
+
+## Required Approvals
+
+- Engineering: @r0mant && @smallinsky
+- Product: @xinding33 || @klizhentas
+- Security: @reedloden || @jentfoo
+
+## What
+
+Extends [automated db user provisioning](0113-automatic-database-users.md) with
+permission management capabilities.
+
+## Why
+
+Database-level permission management is a natural extension of current Teleport
+RBAC capabilities. The described model should integrate seamlessly with TAG,
+providing a future-proof solution.
+
+Database administrators will be able to use Teleport for both user and
+permission management.
+
+## Details
+
+### Database object import rules
+
+Teleport will fetch database objects and store them as resources in the backend,
+using a set of global "database object import rules". At certain points in time,
+the database schema will be read and passed through the import rules. Each
+import rule defines a set of database labels that must match in order for the
+import rule to fire. The rules are processed in order (as defined by the
+`priority` field).
+
+Each import rule defines a set mappings. Individual mapping consists of a set of
+attributes to match, like `name`, `object kind` or `schema`, and a list of
+labels to apply. The expected values for an attribute are specified as a list,
+while the matching uses glob-like semantics, with `*` matching any set of
+characters.
+
+#### Examples
+
+The `widget-prod` import rule applies to databases with `env: prod` label. It
+has the priority 10. If there is a database object that matches the clauses in
+`spec.mappings.match`, the import rule will apply the specified labels to the
+database object (`env: prod`, `product: WidgetMaster3000`).
+
+```yaml
+kind: db_object_import_rule
+version: v1
+metadata:
+  name: rule_widget_prod
+spec:
+  priority: 10
+  db_labels:
+    env: prod
+  mappings:
+    - match:
+        databases:
+          - 'Widget*'
+        db_service_names:
+          - all-things-widget
+        names:
+          - '*sales*'
+        object_kinds:
+          - table
+          - view
+          - procedure
+        protocol_names:
+          - postgres
+        schemas:
+          - widget
+          - sales
+          - public
+          - secret
```

Comment:
I feel we can further simplify this

First, similar to db objects, `db_service_names` and `protocol_names` are "attributes" of the database service. `db_labels` should be enough to match the database service without needing the matching on these attributes.

This is a little repetitive to me:
```
       names:
          - '*sales*'
        object_kinds:
          - table
          - view
          - procedure
```

What you think about something like this instead
```
        # search in databases/schemas, default '*'
        databases:
          - 'Widget*'
        schemas:
          - public

        # import tables/views/procedures, default null
        tables:
          - '*sales*'
        views:
          - '*'
        prodecures:
```
Though it's kind of blurry without the comments what are search scopes vs what are getting imported. Maybe `in_databases/in_schemas` to differentiate

Reply:
Hmm, so basically combine (object kind, name) matchers by defining different branches?

It functionally renders `object kind` an enum, but that isn't necessarily a bad thing.

I agree this is more readable in general.

Regarding the search scope vs import - maybe something like this?

```yaml
    scope: # search in databases/schemas, default '*'
        databases:
          - 'Widget*'
        schemas:
          - public
    import: # import tables/views/procedures, default null
        tables:
          - '*sales*'
        views:
          - '*'
        procedures:
          - '*'
```

Reply:
If splitting scopes from this list, it might be easier to move them outside `mappings`?
```yaml
kind: db_object_import_rule
version: v1
metadata:
  name: rule_widget_prod
spec:
  priority: 10
  db_labels:
     env: prod
  database_names: ['Widget']
  schema_names: ['*']
  mappings:
    - match:
      table_names:
       - 'sales'
      add_labels:
        env: prod
        dept: sales
```

As others pointed out, if to be consistent with okta rules, may want to use like `table_names` and `table_name_regexes`. Any concern with performance using regexes vs simple glob?

Another point from okta rule is each match only allows one import kind:
https://github.com/gravitational/teleport/blob/4f771403dc4177dc26ee0370f7332f3fe54bee0f/api/types/okta.go#L180-L184
The tradeoff is the user may need many `match` blocks but it's clear what each `match` is importing. Might be easier for the implementation too. What do you think?

Reply:
> If splitting scopes from this list, it might be easier to move them outside mappings?

I've experimented with this a bit, and I feel it works better if you can have a set of different `scopes` per each import rule mapping. I'm not entirely sure what the best way to model this is, so I'll keep looking into this.

> (as db_labels is kind of a scope 🤔?) Will database and schema be imported as db_objects in the future?

Both are possible, since these are entities recognised by db engines that we can grant permissions to.

> As others pointed out, if to be consistent with okta rules, may want to use like table_names and table_name_regexes. Any concern with performance using regexes vs simple glob?

Good point regarding naming, this will make it more consistent. No performance penalty: we compile globs to regexes anyway.

> Another point from okta rule is each match only allows one import/object kind:

I'm not sure why this is a restriction for Okta rules, but I think it makes sense to allow multiple entities and kinds.